dll/win32/shlwapi/utils.cpp

```diff
@@ -115,3 +133,99 @@ PathFileExistsDefExtAndAttributesW(
 
     return TRUE;
 }
+
+static inline BOOL
+SHLWAPI_IsBogusHRESULT(HRESULT hr)
+{
+    return (hr == E_FAIL || hr == E_INVALIDARG || hr == E_NOTIMPL);
+}
+
+/*************************************************************************
+ * IShellFolder_GetDisplayNameOf [SHLWAPI.316]
+ */
+EXTERN_C HRESULT WINAPI
+IShellFolder_GetDisplayNameOf(
+    _In_ IShellFolder *psf,
+    _In_ LPCITEMIDLIST pidl,
+    _In_ DWORD uFlags,
+    _Out_ LPSTRRET lpName,
+    _In_ DWORD dwRetryFlags)
+{
+    HRESULT hr;
+
+    TRACE("(%p, %p, 0x%lX, %p, 0x%lX)\n", psf, pidl, uFlags, lpName, dwRetryFlags);
+
+    hr = psf->GetDisplayNameOf(pidl, uFlags, lpName);
+    if (!SHLWAPI_IsBogusHRESULT(hr))
+        return hr;
+
+    dwRetryFlags |= 0x80000000;
+
+    if ((uFlags & SHGDN_FORPARSING) == 0)
+        dwRetryFlags |= SFGDNO_RETRYWITHFORPARSING;
+
+    /* It seems the function is actually retrying here */
+    FIXME("dwRetryFlags: 0x%X\n", dwRetryFlags);
+
+    return hr;
+}
+
+/*************************************************************************
+ * IShellFolder_ParseDisplayName [SHLWAPI.317]
+ */
+EXTERN_C HRESULT WINAPI
+IShellFolder_ParseDisplayName(
+    _In_ IShellFolder *psf,
+    _In_opt_ HWND hwndOwner,
+    _In_opt_ LPBC pbcReserved,
+    _In_ LPOLESTR lpszDisplayName,
+    _Out_opt_ ULONG *pchEaten,
+    _Out_opt_ PIDLIST_RELATIVE *ppidl,
+    _Out_opt_ ULONG *pdwAttributes)
+{
+    ULONG dummy1, dummy2;
+
+    TRACE("(%p, %p, %s, %p, %p, %p)\n", hwndOwner, pbcReserved, lpszDisplayName,
+                                        pchEaten, ppidl, pdwAttributes);
+
+    /* Improve safety */
+    dummy1 = dummy2 = 0;
+
+    if (!pdwAttributes)
+        pdwAttributes = &dummy1;
+
+    if (!pchEaten)
+        pchEaten = &dummy2;
+
+    if (ppidl)
+        *ppidl = NULL;
+
+    return psf->ParseDisplayName(hwndOwner, pbcReserved, lpszDisplayName, pchEaten,
+                                 ppidl, pdwAttributes);
+}
+
+/*************************************************************************
+ * IShellFolder_CompareIDs [SHLWAPI.551]
+ */
+EXTERN_C HRESULT WINAPI
+IShellFolder_CompareIDs(
+    _In_ IShellFolder *psf,
+    _In_ LPARAM lParam,
+    _In_ PCUIDLIST_RELATIVE pidl1,
+    _In_ PCUIDLIST_RELATIVE pidl2)
+{
+    HRESULT hr;
+    TRACE("(%p, %p, %p, %p)\n", psf, lParam, pidl1, pidl2);
+
+    if (lParam & ~(SIZE_T)SHCIDS_COLUMNMASK)
+    {
+        /* Try as IShellFolder2 if possible */
+        hr = psf->QueryInterface(IID_IShellFolder2, (void **)&psf);
+        if (FAILED(hr))
+            lParam = LOWORD(lParam);
```

Comment:
Why does it need to be truncated to the low word? since the CompareIDs function below accept a regular lparam? Otherwise it may be better to do: `lParam &= SHCIDS_COLUMNMASK;` to understand what the underlying purpose really is? (I'm basing the suggestion on the values of the masks you've defined in the IDL file.)

Reply:
> Why does it need to be truncated to the low word?

You are technically not allowed to pass the upper 16-bits to a IShellFolder(v1), only IShellFolder2 is guaranteed to understand (and filter out) the upper 16-bits.